### ISSUE TITLE
[BUG] Error on app startup

### ISSUE THREAD

**Issue by duskull88** [Akylas/OSS-DocumentScanner#337]
### Version

1.12.1 Build 101

### What platform are you using?

Android

### OS Version

Android 14, LineageOS 21. Device is Motorola Moto G32.

### What happened?

On app opening, I get the following error. Scanning is not possible at all, navigating the settings is possible.

![screen1](https://github.com/user-attachments/assets/607e0301-e3b5-4172-a9fb-629e5101dab7)

I installed the app from the IzzyOnDroid repo.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by farfromrefug** [comment]
@duskull88 really sorry about that. Right now the error message is not sufficient for me to help you. In the next version the error message will be explicit. I ll let you know when it is released so that you can test again.
Thanks

**Comment by duskull88** [comment]
Thank you for your reply @farfromrefug . Now I have version 1.13.0 Build 106 and get the following error. Also, I can open the camera screen now, but the capture button to take a photo is constantly blinking, very fast.

![Scr_07-12-2024](https://github.com/user-attachments/assets/35d794a1-8d80-4628-b33d-02242b0da00c)

**Comment by farfromrefug** [comment]
@duskull88 sorry for the latest answer. Is this still an issue in latest versions?

**Comment by duskull88** [comment]
Hello @farfromrefug , yes, the issue still persists. The app version is now 1.14.3 Build 119 from the IzzyOnDroid repo.
I am still using the same device, however, the OS version is now LineageOS 22.1 (Android 15).
I tried to remove the app data and reinstall the app, but issue is still the same.

When I open the app:

![Image](https://github.com/user-attachments/assets/2c90caac-4288-4f9a-83f4-1e349af363a9)

And when I try to scan a document:

![Image](https://github.com/user-attachments/assets/8b0e4ab6-77d6-4b37-bce9-b275527ca515)

**Comment by farfromrefug** [comment]
@duskull88 Ok sorry to hear that. Do you have an sdcard? I am seeing in your screenshot `/storage/e01bf....` This seems to be a weird sdcard path.
I am thinking the app sees `/storage/e01bf....`  as a volume where it can write when it actually cant.

**Comment by duskull88** [comment]
@farfromrefug  Yes, I do use an SD card. I also agree that the patch to it is weird, I am not very up to that what is going on in the Android world lately. But no other app gives me errors about trying to access the memory card. Do you think that I can do something on my end to fix this?

**Comment by farfromrefug** [comment]
@duskull88 no i think something must be wrong with the code i use to "discover" your sdcard path. Thus the app tries to create a file in a folder either not existing or on which it does not have permissions.
What bugs me is that you are on LOS 14 like i am so we should see the same thing :s If you use Material Files or other apps like it what is the true path of your sdcard?
Do you use LOS features like work profiles or things like that (which could induce sandbox environment)?

**Comment by farfromrefug** [comment]
i created a very simple test app which shows the available paths on your device. Could you test it and share a screenshot with me?
[app-release.zip](https://github.com/user-attachments/files/18704547/app-release.zip)

**Comment by duskull88** [comment]
@farfromrefug I do not use work profiles or any other sandboxing features, at least as far as I am aware of.
Thanks for the little test app, here is the screenshot of it:

![Image](https://github.com/user-attachments/assets/2d517de4-7aa8-4547-840b-45d163876ca1)

**Comment by farfromrefug** [comment]
@duskull88 thanks! this confirms i am doing it the right way. But now i still dont understand what s going on. You installed from fdroid right?

Could you try this build? It will send me a bug report. Hopefully i ll enough in the logs to fix it
[app-arm64-v8a-release.zip](https://github.com/user-attachments/files/18737053/app-arm64-v8a-release.zip)

**Comment by duskull88** [comment]
@farfromrefug Thanks for your quick replies and your dedication for this Project!
So I installed the build above, I opened it and nothing different happened. Did it send out some logs to you automatically?

**Comment by farfromrefug** [comment]
@duskull88 i dont think i received it. Can you try to kill the app and restart it? It should force the sending of the bug report

**Comment by duskull88** [comment]
@farfromrefug Perfect, thank you!

**Comment by DJCrashdummy** [comment]
i agree, something weird is going on here... i do get the same error and can somewhat reliable reproduce it:

1. it only appears on a **device with** an **external micro sd card inserted**. - on an other device with no **external** micro sd card inserted (which would be able to take one though), everything works fine... although the setting _"storage"_ within the app is also set to _"sd card"_ (my translations) by default and not changed. 🤨
2. if i try to change the mentioned setting to _"internal storage"_ (to circumvent some suspected permission issues), i get a warning that my data will be moved and the app restarted, followed by an other error as soon as i tap _"OK"_: `java.io.IOException: Permission denied` - and the setting stays at _"sd card"_.
3. to be on the safe side and mention it: the alphanumerical part of the path in the main error message is the **UUID of the sd card**... which is **formatted with an ext4 file system** but certainly not with a FAT or exFAT file system.
4. ...and it seems that this is somehow the main issue: even a file browser app can't access the "Android" folder on the ext4-formatted sd card but the ones on the exFAT-formatted sd card and internal storage without any issue (with Shizuku permissions).

so long story, short: i'm pretty sure this is a **permission issue because of the file system** resp. perhaps their permissions(?)... but somehow it must be possible to workaround this because i know no other app having issues with this. 🤔

PS: this was tested with version 1.16.8 build 131 (from Izzys fdroid repo) on devices with Android 15 resp. LineageOS 22.2.

PPS: the title may be somewhat misleading (i found this bug only because of the key word "sd card" - because i already suspected the issue has to do with it), as i sometimes actually have to perform a scan - which doesn't work - at least once after a reset to invoke the error...

**Comment by farfromrefug** [comment]
@DJCrashdummy Thanks for the detailed report. I think you cant migrate to "internal" because it tries to migrate from sdcard but get a permission issue reading the sdcard...
one solution i think if you want to use the app right now is to:
* remove the sdcard
* clear data of the app and start it
* it will use internal from now on even if you add the sdcard back. (If there is no bug there :s)

Will try to investigate your issue. But really tricky as i can't reproduce it here

**Comment by DJCrashdummy** [comment]
@farfromrefug thanks for your reply!

> I think you cant migrate to "internal" because it tries to migrate from sdcard but get a permission issue reading the sdcard...

yes, as already mentioned, i'm pretty sure it is something like that. -> using the **external** sd card by default is perhaps a somewhat unfortunate decision? 🤔
- users without any external sd card, don't bother at all...
- users who insert an sd card afterwards have to migrate **manually** anyway if they really want to use the external sd card for the app data. ~~_(or not...? - which would be even worse IMHO!)_~~
- users who already have an external sd card inserted are in **danger of loosing their data** when they remove it... and noobs may be surprised by that as they have saved all their personal data from the external sd card but have no clue what the "Android" folder contains (as they actually not allowed access it - without special permissions - on current Android devices no matter the file system). ...and hardly any user thinks about going through every app to eventually migrate app data before removing an external sd card.

> one solution [...] to use the app right now [...]

thanks for the suggestion, i haven't thought about that. yes, this workaround works perfectly fine! 👍 interestingly even without setting the storage manually to "internal storage" while no external sd card is inserted, it is set **automatically** to "internal storage" **after** inserting one. *(but i guess this clever approach is on purpose that nobody has to search for his scanned documents after inserting an external sd card.)*

i'm not even close to an android developer, so bear with me if this is a dumb idea or if i make it sound easier than it is... but perhaps a permission check resp. a check if the needed folder is writable **before** actually **even trying** to use it, especially if is set by default to "sd card" may be a solution or at least an easier workaround?

> [...] i can't reproduce it [...]

you can't reproduce it even with an external sd card formatted as ext4? 🤨 perhaps try to **format** the sd card with an ext4 file system either with android resp. **LinageOS** or on an other machine with an **other OS than android**... i can't remember what i did years ago when formatting the external sd card.
i just can tell you, that it has an **GPT partition table**, an **ext4 file system** and interestingly the **owner and group id** of nearly all files and folders (which are accessible without any issue on android) is `10192` except for "lost+found" and "Android" (with its containing folders/files) it is `root` resp. `0`. and all folders or files - no matter the owner - have the **permissions `700` or `600`**.

so it is in fact a permission issue caused by the file system resp. set owner of the external sd card... but i'm wondering why _"OSS Document Scanner"_ is the only app complaining about that. 🤨

- - -

not sure if it is useful at all, but to mention it just in case: one FOSS app i know which for sure also has such a feature of moving around the app data between the internal and the external sd card is OsmAnd. - and AFAIK it didn't have any issues accessing and using the "Android" folder on an external sd card ever. but don't take my word on that for granted, as it was a looong time ago i used an external sd card for app data and i'm not 100% sure if it was also ext4 formatted back then.

**Comment by farfromrefug** [comment]
@DJCrashdummy they must be using sac. i will try a few things. 
i am not sure i have a sdcard laying around and i really dont want to format mine.
might just take a bit of time

**Comment by DJCrashdummy** [comment]
ok... i was somewhat puzzled because it seemed that you can't reproduce the issue (even with formatting an external micro sd card as ext4).

perhaps for now an easier fix/workaround than debugging the permission issue might be the mentioned permission check resp. not using the external micro sd card by default... what do you think about that, @farfromrefug?

**Comment by farfromrefug** [comment]
@DJCrashdummy no i cant reproduce it but i implemented a check to see if we "can" use the sdcard.
Could you try with latest?

**Comment by farfromrefug** [comment]
@DJCrashdummy i ll close it for now. Let me know if it still happens
